python/pyarrow/pandas-shim.pxi

```diff
@@ -101,8 +101,10 @@ cdef class _PandasAPIShim(object):
                 self._import_pandas(raise_)
             return
 
-        self._tried_importing_pandas = True
-        self._import_pandas(raise_)
+        try:
+            self._import_pandas(raise_)
+        finally:
+            self._tried_importing_pandas = True
```

Comment:
Since we're bothered about race conditions, can we use a lock here? 
```suggestion
        with self._lock:
            if self._tried_importing_pandas:
                try:
                    self._import_pandas(raise_)
                finally:
                    self._tried_importing_pandas = True
```

Reply:
Yes I considered adding a lock, but I was a bit concerned about potential performance implications. May concurrent calls to `_check_import`  become a bottle neck as threads wait to acquire the lock? Thoughts? I will follow your guidance here.

Reply:
In most cases, this method will return from line 102 above, right?

Reply:
> In most cases, this method will return from line 102 above, right?

Ah yes good point, a lock should not hurt performance in those cases. I can go ahead with that approach